pkg/instancetype/instancetype.go

```diff
@@ -510,31 +509,134 @@ func (m *methods) getInstancetypeSpecRevision(revisionName string, namespace str
+	instancetype, err := m.clientset.VirtualMachineClusterInstancetype().Get(context.Background(), name, metav1.GetOptions{})
 	if err != nil {
 		return nil, err
 	}
 	return instancetype, nil
 }
 
-func (m *methods) findClusterInstancetype(vm *virtv1.VirtualMachine) (*instancetypev1alpha1.VirtualMachineClusterInstancetype, error) {
-
-	if vm.Spec.Instancetype == nil {
+func (m *methods) FindPreferredInstancetype(vm *virtv1.VirtualMachine) (*virtv1.InstancetypeMatcher, error) {
+	if vm.Spec.Instancetype != nil || len(vm.Spec.Template.Spec.Volumes) == 0 {
 		return nil, nil
 	}
 
-	instancetype, err := m.clientset.VirtualMachineClusterInstancetype().Get(context.Background(), vm.Spec.Instancetype.Name, metav1.GetOptions{})
-	if err != nil {
-		return nil, err
+	var (
+		preferredInstancetype     string
+		preferredInstancetypeKind string
+		hasAnnotation             bool
+		hasKindAnnotation         bool
+		instancetype              *instancetypev1alpha1.VirtualMachineInstancetype
+		clusterInstancetype       *instancetypev1alpha1.VirtualMachineClusterInstancetype
+		instancetypeSpec          instancetypev1alpha1.VirtualMachineInstancetypeSpec
+	)
+
+	// Kind defaults to cluster as it does when users provide a InstancetypeMatcher within the VirtualMachine
+	preferredInstancetypeKind = apiinstancetype.ClusterSingularResourceName
+
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		if volume.PersistentVolumeClaim != nil {
+			pvc, err := m.clientset.CoreV1().PersistentVolumeClaims(vm.Namespace).Get(context.Background(), volume.PersistentVolumeClaim.ClaimName, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			if preferredInstancetype, hasAnnotation = pvc.Annotations[preferredInstancetypeAnnotation]; hasAnnotation {
+				if _, hasKindAnnotation = pvc.Annotations[preferredInstancetypeKindAnnotation]; hasKindAnnotation {
+					preferredInstancetypeKind = pvc.Annotations[preferredInstancetypeKindAnnotation]
+				}
+
+				// validate the provided type and kind by attempting to fetch
+				switch preferredInstancetypeKind {
+				case apiinstancetype.SingularResourceName, apiinstancetype.PluralResourceName:
+					instancetype, err = m.findInstancetype(preferredInstancetype, vm.Namespace)
+					if err != nil {
+						return nil, err
+					}
+					instancetypeSpec = instancetype.Spec
+				case apiinstancetype.ClusterSingularResourceName, apiinstancetype.ClusterPluralResourceName:
+					clusterInstancetype, err = m.findClusterInstancetype(preferredInstancetype)
+					if err != nil {
+						return nil, err
+					}
+					instancetypeSpec = clusterInstancetype.Spec
+				default:
+					return nil, fmt.Errorf("got unexpected kind in annotation: %s", preferredInstancetypeKind)
+				}
+
+				// If the preferredInstancetype conflicts with the user provided VirtualMachine then ignore it
+				if err = m.checkForInstancetypeConflicts(&instancetypeSpec, &vm.Spec.Template.Spec); err != nil {
+					return nil, err
+				}
```

Comment:
ignoring the conflict could cause unexpected behavior for end users.

Reply:
Yeah at the time I thought this would actually improve the user experience as the fact they provided a VM that conflicts suggests they don't want to use the preferredInstancetype.